### PR TITLE
Support text-generation in InferenceClient

### PR DESCRIPTION
**Pull request by Wauplin**
Original implementation taken from the `text-generation-inference` Python client (see [client library](https://pypi.org/project/text-generation/) and [repo](https://github.com/huggingface/text-generation-inference/tree/main/clients/python)) from @OlivierDehaene . A vast majority of the code comes from there so kudos goes to him :pray: .

Changes compared to the original implementation:
- use `pydantic.dataclasses` instead of `BaseModel`
- default to Python's dataclasses if pydantic is not installed (meaning pydantic/arg validation is recommended but optional)
  - added default values for optional parameters (not needed in `BaseModel` but `dataclasses` yes)
- integrated in `huggingface_hub.InferenceClient`
- added `stream: bool` and `details: bool` in the `text_generation` method instead of having different methods for each use case
- NO asyncio support yet => TODO in a next PR

If model is not served with TGI backend (example: "gpt2"-like models), some parameters are ignored. The client always consider that TGI is enable but default back to a normal call if that's not the case. A warning is triggered for the user + `details=True` is not possible.

Integration is now functional and locally tested.

Docs: [text_generation](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1513/en/package_reference/inference_client#huggingface_hub.InferenceClient.text_generation) and [dataclasses descriptions](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1513/en/package_reference/inference_client#huggingface_hub.inference._text_generation.TextGenerationParameters).
.
**TODO:**
- [x] handle models not served with text-generation-inference backend
- [x] add unit tests
- [x] add documentation / reference

---
## Normal use (with details)

```py
from huggingface_hub import InferenceClient

client = InferenceClient()
output = client.text_generation("\ndef hello_world(name: str)", model="bigcode/starcoder", stream=False, details=True)
print(output)
```

```py
TextGenerationResponse(generated_text=' -> str:\n    return f"Hello {name}"\n\n\ndef test_hello_world():', details=Details(finish_reason=<FinishReason.Length: 'length'>, generated_tokens=20, seed=None, prefill=[], tokens=[Token(id=967, text=' ->', logprob=-0.0009317398, special=False), Token(id=596, text=' str', logprob=-0.25805664, special=False), Token(id=44, text=':', logprob=-0.0005950928, special=False), Token(id=284, text='\n   ', logprob=-0.07080078, special=False), Token(id=442, text=' return', logprob=-0.6455078, special=False), Token(id=296, text=' f', logprob=-0.40722656, special=False), Token(id=20, text='"', logprob=-0.37060547, special=False), Token(id=8279, text='Hello', logprob=-0.11077881, special=False), Token(id=301, text=' {', logprob=-0.67626953, special=False), Token(id=426, text='name', logprob=-0.006881714, special=False), Token(id=3845, text='}"', logprob=-0.72021484, special=False), Token(id=478, text='\n\n', logprob=-0.56689453, special=False), Token(id=203, text='\n', logprob=-0.009246826, special=False), Token(id=589, text='def', logprob=-1.2724609, special=False), Token(id=894, text=' test', logprob=-1.5761719, special=False), Token(id=81, text='_', logprob=-0.012329102, special=False), Token(id=7656, text='hello', logprob=-1.8945312, special=False), Token(id=81, text='_', logprob=-0.071777344, special=False), Token(id=5860, text='world', logprob=-0.10595703, special=False), Token(id=2262, text='():', logprob=-0.7866211, special=False)], best_of_sequences=None))
```

--- 

## Streaming (no details)

```py
for output in client.text_generation(
    "Number between 1 and 1000: 1, 2, ", max_new_tokens=100, model="bigcode/starcoder", stream=True, details=False
):
    print(output, sep="", end="", flush=True)
```
```
3, 4, 5, 6, 7, 8, 9, (...)
```

### CODE REVIEW

File: src/huggingface_hub/inference/_client.py [hunk 1]
```diff
@@ -659,6 +684,213 @@ def summarization(
...
+        Returns:
+            Response: generated response
+        """
+        # NOTE: Text-generation integration is taken from the text-generation-inference project. It has more features
+        # like input/output validation (if Pydantic is installed). See `_text_generation.py` header for more details.
+
+        # Validate parameters
+        parameters = TextGenerationParameters(
+            best_of=best_of,
+            details=True,
```
Comment (OlivierDehaene) [inline src/huggingface_hub/inference/_client.py:864]:
```suggestion
            details=details,
```

File: src/huggingface_hub/inference/_client.py [hunk 2]
```diff
@@ -659,6 +684,213 @@ def summarization(
...
+
+        # Handle errors separately for more precise error messages
+        try:
+            response = self.post(json=asdict(request), model=model, task="text-generation", stream=stream)
+        except HTTPError as e:
+            raise_text_generation_error(e)
+
+        # Parse output
+        if stream:
+            return _stream_text_generation_response(response, details)  # type: ignore
```
Comment (OlivierDehaene) [inline src/huggingface_hub/inference/_client.py:889]:
```suggestion
            return _stream_text_generation_response(response, details or decoder_input_details)  # type: ignore
```

File: src/huggingface_hub/inference/_client.py [hunk 3]
```diff
@@ -659,6 +684,213 @@ def summarization(
...
+        else:
+            output = TextGenerationResponse(**response.json()[0])
+            return output if details else output.generated_text
```
Comment (OlivierDehaene) [inline src/huggingface_hub/inference/_client.py:892]:
```suggestion
            return output if details or decoder_input_details else output.generated_text
```

**Review by OlivierDehaene — COMMENTED** [review]
the backend forces details = True if decoder_input_details = True so maybe it should be the same here.

File: src/huggingface_hub/inference/_client.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/inference/_client.py:864]: Good to know, thanks! I'll make the update

File: src/huggingface_hub/inference/_text_generation.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    do_sample: bool = False
+    # Maximum number of generated tokens
+    max_new_tokens: int = 20
+    # The parameter for repetition penalty. 1.0 means no penalty.
+    # See [this paper](https://arxiv.org/pdf/1909.05858.pdf) for more details.
+    repetition_penalty: Optional[float] = None
+    # Whether to prepend the prompt to the generated text
+    return_full_text: bool = False
+    # Stop generating tokens if a member of `stop_sequences` is generated
+    stop: List[str] = field(default_factory=lambda: [])
```
Comment (StephenHodgson) [inline src/huggingface_hub/inference/_text_generation.py:59]: I noticed we are not validating the stop sequences. I believe the max is 4

Comment (Wauplin) [reply src/huggingface_hub/inference/_text_generation.py:59]: Thanks for the heads up. @OlivierDehaene could you confirm this 4-items limit? I found the same in the [openapi.json](https://github.com/huggingface/text-generation-inference/blob/53aa9194c8c070afd19fa4660305dab2b280adf3/docs/openapi.json#L507) specs but prefer to cross-check with you

Comment (OlivierDehaene) [reply src/huggingface_hub/inference/_text_generation.py:59]: `max_best_of` and `max_stop_sequences` are both parameters that can be modified in TGI. The defaults are 2 and 4 but they can also be turned off.

Comment (Wauplin) [reply src/huggingface_hub/inference/_text_generation.py:59]: Ok so let's not validate them client-side. Thanks for confirming

**Review by LysandreJik — APPROVED** [review]
Impressive conditional type response decided by the values of details/stream! Is this so that pydantic behaves correctly?

I didn't find a guide/example showcasing how you'd use it in practice.  I think it'd be worth adding to the docs as a docstring example.

For example, four small examples showcasing the difference between the `details` and `stream` modes and self-explanatory results about their content. I'm sure a motivated user could go and search for the docs of `TextGenerationResponse` an `Token`, but something like this would get that info straight away in the doc page they'd be looking at:

---

This method has four different return possibilities, according to the value of the `details` and `stream` parameters passed to the method.

With `details` and `stream` as `False`:

```py
>>> from huggingface_hub import InferenceClient
>>> client = InferenceClient()
>>> client.text_generation("What's up")

" with the weather?\nI'm sorry, I am an AI language model and do not have"
```

If `details` is `True` and `stream` is `False`:

```py
>>> client.text_generation("What's up", details=True)

TextGenerationResponse(
    generated_text=" with the weather?\nI'm sorry, I am an AI language model and do not have", 
    details=Details(
        finish_reason=<FinishReason.Length: 'length'>, 
        generated_tokens=20, 
        seed=None, 
        prefill=[InputToken(id=1562, text='What', logprob=None), InputToken(id=18, text="'", logprob=-2.5390625), InputToken(id=94, text='s', logprob=-0.13061523), InputToken(id=510, text=' up', logprob=-4.2382812)], 
        tokens=[Token(id=335, text=' with', logprob=-1.4267578, special=False), Token(id=248, text=' the', logprob=-1.4677734, special=False), Token(id=5015, text=' weather', logprob=-3.15625, special=False), Token(id=42, text='?', logprob=-0.4638672, special=False), Token(id=193, text='\n', logprob=-0.06262207, special=False), Token(id=52, text='I', logprob=-0.22729492, special=False), Token(id=18, text="'", logprob=-0.0769043, special=False), Token(id=88, text='m', logprob=-0.0022010803, special=False), Token(id=6893, text=' sorry', logprob=-0.027435303, special=False), Token(id=23, text=',', logprob=-0.033081055, special=False), Token(id=295, text=' I', logprob=-0.7036133, special=False), Token(id=653, text=' am', logprob=-0.9658203, special=False), Token(id=267, text=' an', logprob=-0.3400879, special=False), Token(id=8317, text=' AI', logprob=-0.052856445, special=False), Token(id=3599, text=' language', logprob=-0.0051193237, special=False), Token(id=2308, text=' model', logprob=-0.0007414818, special=False), Token(id=273, text=' and', logprob=-0.019866943, special=False), Token(id=441, text=' do', logprob=-0.40722656, special=False), Token(id=416, text=' not', logprob=-0.00076293945, special=False), Token(id=413, text=' have', logprob=-0.011177063, special=False)], 
        best_of_sequences=None
    )
)
```

...

---

Overall, it works very well. The tests look good. The cassettes could be offloaded to a dataset repo if you don't want to weigh down the repo, but no strong need to move it out as it's code only.

**Comment by Wauplin** [comment]
Thanks @LysandreJik for the review and suggesting some improvements in the docs! :heart: I've made the requested changes (mainly the return type + adding some examples). [Check it out in the docs](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1513/en/package_reference/inference_client#huggingface_hub.InferenceClient.text_generation).


> Impressive conditional type response decided by the values of details/stream! Is this so that pydantic behaves correctly?

About the return type with the `@overload` type annotation, the main goal is to help with IDE autocompletion / type checks in downstream libraries. It tells any tool what will be the returned object depending on the `details` and `stream` input values. In theory, Pydantic is quite good at validating things with `Union[...]` but I'm not using it here for this.

> The cassettes could be offloaded to a dataset repo if you don't want to weigh down the repo, but no strong need to move it out as it's code only.

Yes that could be a possibility but for now I'd prefer not to, just for the sake of keeping it simple. The main advantage of having it directly in the repo is that updating the cassette is as simple as running pytest with `--vcr-record=all`. No need to re-upload the fixed cassettes which makes the process simpler. I might do it in a future PR if it become too heavy.

**Pull request merged**